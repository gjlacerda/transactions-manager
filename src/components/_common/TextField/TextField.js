import React, { Component } from 'react'
import styled from 'styled-components'
import { colorApp, colorBlack } from 'utils/color'
import { fontSize, fontWeight } from 'utils/font'
import { opacityLight } from 'utils/opacity'
import { transitionInput } from 'utils/transition'

const Container = styled.div`
  display: flex;
  align-items: flex-end;
  position: relative;
  height: 48px;
  opacity: ${opacityLight};
  transition: ${transitionInput('opacity')};

  &:hover {
    opacity: 1;
  }

  &.selected {
    opacity: 1;

    label {
      transform: translate(0, -25px) scale(0.8);
      color: ${colorApp};
    }
  }
`

const Label = styled.label`
  position: absolute;
  top: 24px;
  left: 0;
  color: ${colorBlack};
  pointer-events: none;
  font-size: ${fontSize.medium};
  font-weight: ${fontWeight.thin};
  transition: transform 0.1s cubic-bezier(0, 0, 0.2, 1);
  transform-origin: left center;
`

const InputText = styled.input`
  background: none;
  border: none;
  border-radius: 0;
  border-bottom: 1px solid ${colorBlack};
  outline: none;
  padding: 0 0 6px 0;
  width: ${props => props.fullWidth ? '100%' : (props.width || 'auto')};
`

class TextField extends Component {
  state = {}

  componentDidMount() {
    const { focus } = this.props
    if (focus) {
      this.input.focus()
    }
  }

  handleFocus = focus => () => {
    this.setState({
      className: 'selected',
    })
    if (focus) {
      focus()
    }
  }

  handleBlur = (event) => {
    if (event.target.value) {
      return
    }
    this.setState({
      className: '',
    })
  }

  render() {
    const {
      label,
      fullWidth,
      width,
      type,
      pattern,
      maxLength,
      id,
      inputmode,
      onChange,
      onFocus,
    } = this.props
    const { className } = this.state

    return (
      <Container className={className}>
        <InputText
          type={type || 'text'}
          step="0.01"
          onFocus={this.handleFocus(onFocus)}
          onBlur={this.handleBlur}
          fullWidth={fullWidth}
          width={width}
          pattern={pattern}
          maxLength={maxLength}
          id={id}
          inputmode={inputmode || 'none'}
          ref={input => this.input = input}
          onChange={onChange}
        />
        <Label>
          {label}
        </Label>
      </Container>
    )
  }
}

export default TextField
