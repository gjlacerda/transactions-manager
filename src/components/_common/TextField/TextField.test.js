import React from 'react'
import { mount } from 'enzyme'
import renderer from 'react-test-renderer'
import 'jest-styled-components'
import TextField from './TextField'

describe('TextField component', () => {
  it('deve ser igual ao snapshot', () => {
    const tree = renderer.create(
      <TextField
        label="Teste"
        fullWidth
      />,
    ).toJSON()
    expect(tree).toMatchSnapshot()
  })

  it('deve chamar a função onChange passada por parâmetro', () => {
    const spy = jest.fn()
    const wrapper = mount(<TextField onChange={spy} />)
    wrapper.find('input').simulate('change', { target: { value: 'teste' } })
    expect(spy).toHaveBeenCalled()
  })
})
