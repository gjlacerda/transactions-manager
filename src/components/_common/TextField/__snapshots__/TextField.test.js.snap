// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TextField component deve ser igual ao snapshot 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: flex-end;
  -webkit-box-align: flex-end;
  -ms-flex-align: flex-end;
  align-items: flex-end;
  position: relative;
  height: 48px;
  opacity: 0.8;
  -webkit-transition: opacity 0.1s ease-in-out;
  transition: opacity 0.1s ease-in-out;
}

.c0:hover {
  opacity: 1;
}

.c0.selected {
  opacity: 1;
}

.c0.selected label {
  -webkit-transform: translate(0,-25px) scale(0.8);
  -ms-transform: translate(0,-25px) scale(0.8);
  transform: translate(0,-25px) scale(0.8);
  color: #1976d2;
}

.c2 {
  position: absolute;
  top: 24px;
  left: 0;
  color: #555;
  pointer-events: none;
  -webkit-transition: -webkit-transform 0.1s cubic-bezier(0,0,0.2,1);
  -webkit-transition: transform 0.1s cubic-bezier(0,0,0.2,1);
  transition: transform 0.1s cubic-bezier(0,0,0.2,1);
  -webkit-transform-origin: left center;
  -ms-transform-origin: left center;
  transform-origin: left center;
}

.c1 {
  background: none;
  border: none;
  border-radius: 0;
  border-bottom: 1px solid #555;
  outline: none;
  padding: 0 0 6px 0;
  width: 100%;
}

<div
  className="c0"
>
  <input
    className="c1"
    onBlur={[Function]}
    onFocus={[Function]}
    step="0.01"
    type="text"
  />
  <label
    className="c2"
  >
    Teste
  </label>
</div>
`;
